fedimint-client/src/sm/executor.rs

```diff
@@ -0,0 +1,614 @@
+        }
+    }
+
+    impl Context for MockContext {}
+
+    impl ClientModule for MockClientModule {
+        type StateMachineContext = MockContext;
+        type States = MockStateMachine;
+
+        fn decoder(&self) -> Decoder {
+            let mut decoder_builder = Decoder::builder();
+            decoder_builder.with_decodable_type::<MockStateMachine>();
+            decoder_builder.build()
+        }
+
+        fn context(&self) -> Self::StateMachineContext {
+            MockContext {
+                broadcast: self.0.clone(),
+            }
+        }
+    }
+
+    async fn get_executor(tg: &mut TaskGroup) -> (Executor, Sender<u64>, Database) {
+        let (broadcast, _) = tokio::sync::broadcast::channel(10);
+        let module = MockClientModule(broadcast.clone());
+
+        let decoders = ModuleDecoderRegistry::new(vec![(42, module.decoder())]);
+        let db = Database::new(MemDatabase::new(), decoders);
+
+        let mut executor_builder = Executor::builder();
+        executor_builder.with_module::<MockClientModule>(42, module);
+        let executor = executor_builder.build(tg, db.clone(), Arc::new(())).await;
+
+        info!("Initialized test executor");
+        (executor, broadcast, db)
+    }
+
+    #[tokio::test]
+    #[tracing_test::traced_test]
+    async fn test_executor() {
+        const MOCK_INSTANCE_1: ModuleInstanceId = 42;
+        const MOCK_INSTANCE_2: ModuleInstanceId = 21;
+
+        let mut task_group = TaskGroup::new();
+        let (executor, sender, _db) = get_executor(&mut task_group).await;
+        executor
+            .add_state_machines(vec![DynState::from_typed(
+                MOCK_INSTANCE_1,
+                MockStateMachine::Start,
+            )])
+            .await
+            .unwrap();
+
+        assert!(
+            executor
+                .add_state_machines(vec![DynState::from_typed(
+                    MOCK_INSTANCE_1,
+                    MockStateMachine::Start
+                )])
+                .await
+                .is_err(),
+            "Running the same state machine a second time should fail"
+        );
+
+        assert!(
+            executor
+                .contains_active_state(MOCK_INSTANCE_1, MockStateMachine::Start)
+                .await,
+            "State was written to DB and waits for broadcast"
+        );
+        assert!(
+            !executor
+                .contains_active_state(MOCK_INSTANCE_2, MockStateMachine::Start)
+                .await,
+            "Instance separation works"
+        );
+
+        // TODO build await fn+timeout or allow manual driving of executor
+        tokio::time::sleep(Duration::from_secs(1)).await;
+        sender.send(0).unwrap();
+        tokio::time::sleep(Duration::from_secs(2)).await;
+
+        assert!(
+            executor
+                .contains_inactive_state(MOCK_INSTANCE_1, MockStateMachine::Final)
+                .await,
+            "State was written to DB and waits for broadcast"
```

Comment:
Are the custom messages in `assert!` supposed to describe the success or failure of the assertion? These describe success, but I would expect the message to describe the failure since that's when it shows up in test logs ...

Reply:
Good question, opinions @dpc?

Reply:
That would mean the current version is ok? It describes why it shouldn't panic/what we assert.

Reply:
I guess. Though I would consider doing it the longer way: "foobar shouldn't fail because the state blablablah".